### PR TITLE
Do some renaming for consistency and update documentation for new SQL types

### PR DESCRIPTION
**Pull request by beikov**
Rename `MapKeyCustomCompositeType` to `MapKeyCompositeType` and `MapKeyCustomType` to `MapKeyType` i.e. remove the `Custom` part from the name to be consistent with other annotations.

### CODE REVIEW

**Comment by gavinking** [comment]
@beikov 

I'm against adding global configuration parameters that affect ORM mappings.

IMO, `hibernate.type.preferred_instant_jdbc_type` is a bad addition, and is a good example of the sort of thing we've been actively trying to move away from.

File: hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java [hunk 1]
```diff
@@ -144,9 +144,9 @@ protected String columnType(int sqlTypeCode) {
 				// Note that 31 is the maximum precision DB2 supports
 				return columnType( DECIMAL );
 			case BLOB:
-				return "blob($l)";
+				return "blob";
 			case CLOB:
-				return "clob($l)";
+				return "clob";
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: @beikov Why did you make this change.

The commit message doesn't explain why you did this.

Please at least separate it into its own commit.

**Comment by gavinking** [comment]
Wait, you already merged this?!

**Comment by beikov** [comment]
>I'm against adding global configuration parameters that affect ORM mappings.
>IMO, hibernate.type.preferred_instant_jdbc_type is a bad addition, and is a good example of the sort of thing we've been actively trying to move away from.

I guess we can discuss this on Zulip later, but it looks like you already did with Steve. We can certainly switch to a more generic variant for configuring the JDBC/SQL types per Java type, though I think it only makes sense to override the preferred types for certain Java types which can be meaningfully represented through different JDBC types.

File: hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java (hunk 1 above)
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: `DB2Dialect`, `DerbyDialect` and `MimerSQLDialect` were the only dialects that specified a length parameter, even though I'm pretty sure other dialects also allow specifying a length. In the end, the length is not really helpful anyway IMO, but the real problem is that we don't have a default length or a way to determine the LOB max size and we ended up with an unresolved variable in a test that I added. So I removed that to be consistent with the other dialects and fix the test I added.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: > but the real problem is that we don't have a default length or a way to determine the LOB max size and we ended up with an unresolved variable in a test that I added. 

OK, that's a big surprise to me, and I don't know how the existing tests didn't pick it up. But it seems like we should fix _that_ problem, not remove the ability to specify the length of a `blob` in DB2.

Because with the change you've made, if I'm not mistaken, it seems that any length the user explicitly specifies via the `@Length` annotation will now be ignored!

> `DB2Dialect`, `DerbyDialect` and `MimerSQLDialect` were the only dialects that specified a length parameter

Sure, this is all very platform-specific stuff. But if the platform _supports_ a length parameter then we shouldn't just ignore it.

So this looks like the wrong "fix" for the problem you encountered.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: >OK, that's a big surprise to me, and I don't know how the existing tests didn't pick it up. But it seems like we should fix that problem, not remove the ability to specify the length of a blob in DB2.
>Because with the change you've made, if I'm not mistaken, it seems that any length the user explicitly specifies via the @Column annotation will now be ignored!

How about we support length in a unified way across all databases in a follow up fix? We could generate a check constraint to assert the length limit if the type doesn't support specifying capacity.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: Well I think some databases (Postgres, if I'm not mistaken) simply don't support a maximum length. But  I don't recall any details or this.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: For MySQL it surely already works correctly.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: >Well I think some databases (Postgres, if I'm not mistaken) simply don't support a maximum length. But I don't recall any details or this.

You are right, but I think these databases usually support check constraints, so we could enforce a length here if we think it is reasonable to do so.

>For MySQL it surely already works correctly.

Not sure what you mean. AFAICS MySQL has no possibility to specify the length for LOB types.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: Let's continue that discussion on this issue I just created: https://hibernate.atlassian.net/browse/HHH-15155

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: > Not sure what you mean. AFAICS MySQL has no possibility to specify the length for LOB types.

It has `blob`, `mediumblob`, `longblob`, etc. I'm sure we already pick the right one based on the specified `length`.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:149]: Right, but there is no length variable in the DDL type "name"/pattern that can specify exact capacity.
